packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -43,19 +79,35 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
        * TODO: Fix the types for the middleware
        */
       strapi.documents.use(async (context, next) => {
-        // @ts-expect-error ContentType is not typed correctly on the context
-        const contentTypeUid = context.contentType.uid;
-        const params = context.args.at(-1) as any;
         // Ignore actions that don't mutate documents
         if (!['create', 'update', 'publish', 'unpublish'].includes(context.action)) {
           return next(context);
         }
-
+        // @ts-expect-error ContentType is not typed correctly on the context
+        const contentTypeUid = context.contentType.uid;
         // Ignore content types not created by the user
         if (!contentTypeUid.startsWith('api::')) {
           return next(context);
         }
 
+        const result = (await next(context)) as any;
+        /**
+         * Publish actions store the created data on the result's versions array.
+         * Otherwise, for create and update, the data is the result.
+         *
+         * TODO: Fix unpublish which is broken since the data that was unpublished is not passed to the middleware
```

Comment:
I'm fine with this for now. I just noticed that clicking "discard changes" in the UI doesn't create any history versions. Is that related, or fixable now?

Reply:
Good thinking I haven't tried that yet, I'll take a quick look before merging so I can create a bug ticket for that and unpublish

Reply:
Sorry didn't read this part "or fixable now?", 🤣 I'll take a look

Reply:
I can fix that here :)